# Homework Assignment 1

**Course:** COSC 073 - Computational Photography  
**Term:** Spring 2026  
**Author:** Kasuti Makau

In [25]:
import numpy, scipy, skimage, matplotlib

## 1.1 Implement a Basic Image Processing Pipeline

### 1.1a RAW Image Conversion (5 points)

**Prerequisite:** Install `dcraw` since RAW files cannot be directly read by `skimage`.
```bash
brew install dcraw
```

#### Step 1: Extract RAW metadata
Run this command to get the black, white and scale values:

In [26]:
!dcraw -4 -d -v -W -T data/Thayer.CR2

Loading Canon EOS 2000D image from data/Thayer.CR2 ...
Scaling with darkness 2044, saturation 16383, and
multipliers 2.293742 1.000000 1.368654 1.000000
Building histograms...
Writing data to data/Thayer.tiff ...


**Extracted values:**

| Parameter | Value |
|-----------|-------|
| Black     | 2044  |
| White     | 16383 |
| R Scale   | 2.293742 |
| G Scale   | 1.000000 |
| B Scale   | 1.368654 |

#### Step 2: Generate the correct TIFF file

**Important:** Delete the `.tiff` file created in Step 1, then run dcraw with different flags:

> **Note:** The `-D` flag (uppercase) outputs the raw sensor data without any processing, while `-d` (lowercase) applies basic scaling. We need the unprocessed data for our pipeline.

In [27]:
!rm data/Thayer.tiff 

In [28]:
!dcraw -4 -D -T data/Thayer.CR2

## 1.1b Python Initials (5 points)
Using [`skimage.io.imread`](https://scikit-image.org/docs/stable/api/skimage.io.html#skimage.io.imread) to load the image as a numpy array, then
  [`ndarray.astype`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html) to convert to double precision.

In [29]:
from skimage.io import imread

img = imread('data/Thayer.tiff')

bits = img.dtype.itemsize * 8
height, width = img.shape
img_double = img.astype('float64')

print(f'{bits} bits per channel')
print(f'{width} x {height} pixels')
print(f'dtype (double): {img_double.dtype}')

16 bits per channel
6020 x 4016 pixels
dtype (double): float64
